rts/motoko-rts/src/stabilization/stable_memory_stream.rs

```diff
@@ -0,0 +1,135 @@
+//! Streamed read/write access to stable memory.
+//! Supporting Cheney's to-space in stable memory.
+
+use core::mem::{size_of, MaybeUninit};
+
+use crate::stable_mem::{ic0_stable64_read, ic0_stable64_write};
+
+use super::grant_stable_space;
+
+/// Streamed reader/writer on stable memory.
+/// Used for the to-space during stabilization.
+///
+/// The memory supports two location-independent streams:
+/// * Streamed reading and updating, used for scanning and patching pointers.
+/// * Streamed writing, used for allocating new objects.
+pub struct StableMemoryStream {
+    /// The pointers used in the serialized stable memory layout are
+    /// relative to this start address of the to-space.
+    base_address: u64,
+    /// Used for reading and updating.
+    scan_address: u64,
+    /// Used for writing.
+    free_address: u64,
+}
+
+pub trait ScanStream {
+    // Determines whether the stream has reached the end.
+    fn scan_completed(&self) -> bool;
+    // Read a value from the stream.
+    fn read<T>(&mut self) -> T;
+    // Read raw data from the stream.
+    fn raw_read(&mut self, data_address: usize, length: usize);
+    // Go back in the stream.
+    fn rewind(&mut self, length: usize);
+    // Skip data in the stream.
+    fn skip(&mut self, length: usize);
+    // Overwrite the value right before the stream position.
+    fn update<T>(&mut self, value: &T);
+    // Overwrite raw data right before the stream position.
+    fn raw_update(&mut self, data_address: usize, length: usize);
+}
+
+pub trait WriteStream {
+    // Append a value at the stream end.
+    fn write<T>(&mut self, value: &T);
+    // Append raw data at the stream end.
+    fn raw_write(&mut self, data_address: usize, length: usize);
+}
+
+impl StableMemoryStream {
+    pub fn open(start_address: u64) -> StableMemoryStream {
+        StableMemoryStream {
+            base_address: start_address,
+            scan_address: start_address,
+            free_address: start_address,
+        }
+    }
+
+    pub fn close(&mut self) {
+        debug_assert!(self.scan_address <= self.free_address);
+    }
+
+    /// Size of this memory space. Also serves to determine the addresses of a subsequently
+    /// written object in the serialized stable memory format.
+    pub fn written_length(&self) -> u64 {
+        debug_assert!(self.base_address <= self.free_address);
+        self.free_address - self.base_address
+    }
+}
+
+impl ScanStream for StableMemoryStream {
+    fn scan_completed(&self) -> bool {
+        debug_assert!(self.scan_address <= self.free_address);
+        self.scan_address == self.free_address
+    }
+
+    fn read<T>(&mut self) -> T {
```

Comment:
Good point. I will analyze this. Maybe there is an optimization potential.

Reply:
**Update**: I checked it for bulk returrn types and fortunately, the Rust compiler optimzes this by using an output parameter instead of copying the result.